SQLQuery_3(homework).sql

```diff
@@ -0,0 +1,303 @@
+FROM Animals a
+INNER JOIN ZooInstitutions zoo ON a.zooId = zoo.Id
+GROUP BY Name, TypeOfSpecies;
+
+--the number of visitors per day
+
+SELECT v.TypeOfDay 'The day of every week', zoo.Name 'Zoo Name', COUNT(v.Id) 'Number of visitors'
+FROM Visitors v
+INNER JOIN ZooInstitutions zoo ON zoo.Id = v.ZooId
+GROUP BY Name, TypeOfDay;
+
+--the ticket price and validity order by start date
+
+SELECT zoo.Name, t.Price, t.StartDate, t.EndDate
+FROM Tickets t
+INNER JOIN ZooInstitutions zoo ON zoo.Id = t.ZooId
+ORDER BY t.StartDate;
+
+--details about the caretakers order desc by salary
+
+SELECT zoo.Name 'Zoo Name', ct.FirstName, ct.LastName, ct.Salary
+FROM Caretakers ct
+INNER JOIN ZooInstitutions zoo ON zoo.Id = ct.ZooId
+ORDER BY ct.Salary DESC;
+
+--average salary by zoo
+
+SELECT zoo.Name 'Zoo Name', AVG(ct.Salary) 'Average salaries'
+FROM Caretakers ct 
+LEFT JOIN ZooInstitutions zoo ON ct.ZooId = zoo.Id
+GROUP BY Name;
+
+--details about the caretakers and the total work areas they take care
+
+SELECT ct.FirstName, ct.LastName, COUNT(a.TypeOfSpecies) 'Number of areas'
+FROM Caretakers_Animals ca
+    INNER JOIN Caretakers ct ON ct.Id = ca.Caretaker
+    INNER JOIN Animals a ON a.TypeOfSpecies = ca.Animal
+GROUP BY FirstName, LastName;
+
+--details about the caretakers and their work area 
+
+SELECT ct.FirstName, ct.LastName, a.TypeOfSpecies 'Type of work area'
+FROM Caretakers_Animals ca
+    INNER JOIN Caretakers ct ON ct.Id = ca.Caretaker
+    INNER JOIN Animals a ON a.TypeOfSpecies = ca.Animal
+GROUP BY FirstName, LastName, TypeOfSpecies;
+
+-- view with caretakers from zoo
+
+GO
+CREATE VIEW ZooCaretakers
+AS
+SELECT zoo.Name 'Zoo Name', ct.FirstName, ct.LastName
+FROM Caretakers ct JOIN ZooInstitutions zoo
+ON ct.ZooId = zoo.Id;
+GO
+
+ SELECT *
+ FROM ZooCaretakers;
+
+ --view total number of tickets by type from a view
+
+ GO 
+ CREATE VIEW TotalTickets
+ AS
+ SELECT COUNT(t.Id) 'Total number o ticket', t.TypeOfTicket 'Type of ticket'
+ FROM Tickets t
+ GROUP BY TypeOfTicket;
+ GO
+
+ SELECT *
+ FROM TotalTickets;
+
+--create procedure to add a number of existing species
+
+GO
+CREATE PROCEDURE AddExistingSpecies(
+    @typeOfSpecies VARCHAR(100),
+    @numberOfSpecies INT
+)
+AS
+BEGIN
+    UPDATE Animals
+    SET NumberOfSpecies = @numberOfSpecies
+    WHERE TypeOfSpecies = @typeOfSpecies;
+END;
+GO
+
+EXEC AddExistingSpecies 'Lion', 3 ;
+
+--create function that return smallest salary by zoo id
+
+GO
+CREATE FUNCTION dbo.GetSmallestSalaryByZooId(
+    @inputId INT
+)
+RETURNS MONEY
+AS
+BEGIN
+    DECLARE @smallestSalary MONEY;
+    SET @smallestSalary = 
+        (SELECT MIN(Salary) FROM Caretakers c JOIN ZooInstitutions ON c.ZooId = ZooInstitutions.Id);
+    RETURN @smallestSalary;
+END;
+GO
+
+SELECT dbo.GetSmallestSalaryByZooId(1) 'Smallest Salary';
+
+-- index
+
+CREATE INDEX TypeOfSpeciesAndNumbers
+ON Animals(TypeOfSpecies,NumberOfSpecies);
+
+SELECT * 
+FROM Animals
+WHERE TypeOfSpecies = 'Crocodile';
```

Comment:
For some reason I cannot add a comment on the png file so I'm adding it here.
By looking at the diagram I can see that there are some redundant relationships that can be removed. As an example: the visitors and the tickets are both linked to the zoo but it would have been enough to link just one of them, let say ticket linked to zoo, and you can still get the information about visitor and zoo by going through tickets.